src/qtism/data/storage/xml/Utils.php

```diff
@@ -353,11 +355,40 @@ public static function valueAsString($value, $encode = true): string
             return $value === true ? 'true' : 'false';
         }
         if ($encode) {
-            return htmlspecialchars((string)$value, ENT_XML1, 'UTF-8');
+            return self::escapeNonCharacterRange((string)$value);
         }
         return (string)$value;
     }
 
+    public static function escapeNonCharacterRange(string $string): string
+    {
+        // Define the XML escape sequences
+        $symbolMap = [
+            '"'  => '&quot;',
+            "'"  => '&apos;',
+            '&'  => '&amp;',
+            '<'  => '&lt;',
+            '>'  => '&gt;',
+            "\t" => '&#x9;',
+            "\r" => '&#xD;',
+            "\x08" => '&#xFFFD;',
+        ];
+
+        // generate replace map
+        for ($char = 0x20; $char <= 0xDF77; $char++) {
```

Comment:
Wait. Looping through so many chars to build the replacement map, and then applying the replacement map on the string is definitely _slower_ than parsing the string char by char as I initially suggested.

I advise to stick to the initial suggestion.